### PR TITLE
Better node placing

### PR DESCRIPTION
**Pull request by ghost**

### CODE REVIEW

File: src/logic/graphMapping.ts [hunk 1]
```diff
@@ -1,36 +1,121 @@
...
   Edge,
   MappedGraph,
   GraphIndex,
   DisassembledGraph,
+  Position,
 } from './commonInterfaces';
+const rand = require('random-seed');
+
+const SWITCH_ALGORITHM_ABOVE_X_NODES = 10;
+const CIRCLE_RADIUS = 40;
```
Comment (songofroland) [inline src/logic/graphMapping.ts:13]: CIRCLE_RADIUS_PX

File: src/logic/graphMapping.ts [hunk 2]
```diff
@@ -1,36 +1,121 @@
...
-    y: Math.random() * 100,
-  }));
+  [Array<MappedNode>, Array<MappedEdge>] {
+
+  const mappedNodes = placeNodes(nodes, edges);
   const mappedEdges = edges.map((edge: Edge, i) => {
     const defaultNode = { x: 0, y: 0 };
-    const node1 = mappedNodes.find((e) => e.id === edge.from) ?? defaultNode;
-    const node2 = mappedNodes.find((e) => e.id === edge.to) ?? defaultNode;
+    const nodeFrom = mappedNodes.find((_, i) => i === edge.from) ?? defaultNode;
```
Comment (songofroland) [inline src/logic/graphMapping.ts:26]: What is `i`?

File: src/logic/graphMapping.ts [hunk 3]
```diff
@@ -1,36 +1,121 @@
...
+
+function placeOnCircle(nodes: Array<number>, edges: Array<Edge>): Array<MappedNode> {
+  const angleStep = 360 / nodes.length;
+  const center = { x: 50, y: 50 };
+  return nodes.map((node: number) =>
+    getPositionOnCircle(node * angleStep, CIRCLE_RADIUS, center));
+}
+
+/**
+ * It won't place nodes too close together. RNG's seed is graph for consistency.
```
Comment (songofroland) [inline src/logic/graphMapping.ts:52]: _RNG's seed is graph for consistency._ does not tell me anything. If u like doccoments then 
_The generated positions will always be the same for the same graph_  would be way clearer(if I guess right)

File: src/logic/graphMapping.ts [hunk 4]
```diff
@@ -1,36 +1,121 @@
...
+ */
+function placeRandomly(nodes: Array<number>, edges: Array<Edge>): Array<MappedNode> {
+  const seed = nodes.toString() + edges.toString();
+  const gen = typedGenerator(seed);
+  const dissallowedRadius = getDissallowedRadiusBasedOnDensity(nodes.length);
+  const positions = positionGenerator(gen, dissallowedRadius);
+
+  return nodes.map(_ => positions.next().value);
```
Comment (songofroland) [inline src/logic/graphMapping.ts:60]: All the logic for placing randomly / on circle should be grouped in a class or a separate module

File: src/logic/graphMapping.ts [hunk 5]
```diff
@@ -1,36 +1,121 @@
...
+    };
+    if (isAllowed(position, usedPositions, disallowedRadius)) {
+      usedPositions.push(position);
+      yield position;
+    }
+  }
+}
+
+function getDissallowedRadiusBasedOnDensity(numberOfNodes: number): number {
+  // This should do the job just fine. Might change later.
```
Comment (songofroland) [inline src/logic/graphMapping.ts:99]: redundant comment

File: src/logic/graphMapping.ts [hunk 6]
```diff
@@ -1,36 +1,121 @@
...
   MappedNode,
   Edge,
   MappedGraph,
   GraphIndex,
   DisassembledGraph,
+  Position,
 } from './commonInterfaces';
+const rand = require('random-seed');
+
+const SWITCH_ALGORITHM_ABOVE_X_NODES = 10;
```
Comment (songofroland) [inline src/logic/graphMapping.ts:12]: PLACE_RANDOMLY_ABOVE_NODES_NUMBER

**Review by songofroland — CHANGES_REQUESTED** [review]

File: src/logic/graphMapping.ts [hunk 7]
```diff
@@ -1,36 +1,121 @@
...
+  const seed = nodes.toString() + edges.toString();
+  const gen = typedGenerator(seed);
+  const dissallowedRadius = getDissallowedRadiusBasedOnDensity(nodes.length);
+  const positions = positionGenerator(gen, dissallowedRadius);
+
+  return nodes.map(_ => positions.next().value);
+}
+
+/**
+ * Uses algorithm (x, y) = (xO + r * cos(theta), yO + r * sin(theta))
```
Comment (songofroland) [inline src/logic/graphMapping.ts:64]: I don't need to know this here. U can put it to line 67

File: src/logic/graphMapping.ts (hunk 1 above)
Comment (wilmhit) [reply src/logic/graphMapping.ts:13]: Its percentage

File: src/logic/graphMapping.ts (hunk 2 above)
Comment (wilmhit) [reply src/logic/graphMapping.ts:26]: iterator

File: src/logic/graphMapping.ts (hunk 1 above)
Comment (songofroland) [reply src/logic/graphMapping.ts:13]: then CIRCLE_RADIUS_PERCENT

File: src/logic/graphMapping.ts (hunk 2 above)
Comment (songofroland) [reply src/logic/graphMapping.ts:26]: Then find a meaningful name for this

File: src/logic/graphMapping.ts (hunk 1 above)
Comment (wilmhit) [reply src/logic/graphMapping.ts:13]: its the same unit everything is measured in.

Comment (songofroland) [reply src/logic/graphMapping.ts:13]: If I came back to this repo after a long piece of time I would scratch my head add need to look through the repo which unit do we use, either add a comment or  add a `PERCENT`

File: src/logic/graphMapping.ts (hunk 2 above)
Comment (songofroland) [reply src/logic/graphMapping.ts:33]: If `edge.from` depends on some arbitrary iteration the something must be wrong

**Review by songofroland — CHANGES_REQUESTED** [review]

File: src/logic/graphMapping.ts (hunk 2 above)
Comment (wilmhit) [reply src/logic/graphMapping.ts:33]: No, list is ordered. `i` is meaningful name by convention.

Comment (wilmhit) [reply src/logic/graphMapping.ts:33]: It actually stands for "index", internet says.

Comment (songofroland) [reply src/logic/graphMapping.ts:33]: Using `e.id` was way better then

Comment (songofroland) [reply src/logic/graphMapping.ts:33]: or rename `i` -> `id` if it is the same

Comment (wilmhit) [reply src/logic/graphMapping.ts:33]: Noone uses `id` but will change

**Review by songofroland — APPROVED** [review]

**Pull request merged**